packages/proximity/src/map-custom/implementation/NonEmpty.ts

```diff
@@ -0,0 +1,188 @@
+import type { HashMap } from '@rimbu/hashed/map';
+import type { Token } from '@rimbu/base';
+import type { RMapBase } from '@rimbu/collection-types/map-custom';
+import {
+  RelatedTo,
+  ArrayNonEmpty,
+  OptLazy,
+  TraverseState,
+  ToJSON,
+  OptLazyOr,
+} from '@rimbu/common';
+import type { Stream, StreamSource, FastIterator } from '@rimbu/stream';
+import { wrapHashMap } from '../wrapping';
+import type { ProximityMap } from '../../map';
+import { getValueWithNearestKey } from '../../common';
+
+const toStringBeginning = /^[^(]+/;
+
+export class ProximityMapNonEmpty<K, V> implements ProximityMap.NonEmpty<K, V> {
+  readonly isEmpty = false;
+
+  constructor(
+    readonly context: ProximityMap.Context<K>,
+    private readonly internalMap: HashMap.NonEmpty<K, V>
+  ) {}
+
+  private plugInternalMap(
+    newInternalMap: HashMap.NonEmpty<K, V>
+  ): ProximityMap.NonEmpty<K, V>;
+  private plugInternalMap(newInternalMap: HashMap<K, V>): ProximityMap<K, V>;
+  private plugInternalMap(newInternalMap: HashMap<K, V>): ProximityMap<K, V> {
+    if (newInternalMap == this.internalMap) {
+      return this;
+    }
+
+    return wrapHashMap(this.context, newInternalMap);
+  }
+
+  get size(): number {
+    return this.internalMap.size;
+  }
+
+  stream(): Stream.NonEmpty<readonly [K, V]> {
+    return this.internalMap.stream();
+  }
+
+  addEntries(
+    entries: StreamSource<readonly [K, V]>
+  ): ProximityMap.NonEmpty<K, V> {
+    return this.plugInternalMap(this.internalMap.addEntries(entries));
+  }
+
+  updateAt<UK = K>(
+    key: RelatedTo<K, UK>,
+    update: RMapBase.Update<V>
+  ): ProximityMap.NonEmpty<K, V> {
+    return this.plugInternalMap(this.internalMap.updateAt(key, update));
+  }
+
+  nonEmpty(): true {
+    return true;
+  }
+
+  assumeNonEmpty(): this {
+    return this;
+  }
+
+  asNormal(): ProximityMap<K, V> {
+    return this;
+  }
+
+  streamKeys(): Stream.NonEmpty<K> {
+    return this.internalMap.streamKeys();
+  }
+
+  streamValues(): Stream.NonEmpty<V> {
+    return this.internalMap.streamValues();
+  }
+
+  mapValues<V2>(
+    mapFun: (value: V, key: K) => V2
+  ): ProximityMap.NonEmpty<K, V2> {
+    return new ProximityMapNonEmpty(
+      this.context,
+      this.internalMap.mapValues(mapFun)
+    );
+  }
+
+  toArray(): ArrayNonEmpty<readonly [K, V]> {
+    return this.internalMap.toArray();
+  }
+
+  /**
+   * Applies `getValueWithNearestKey()` to its entries
+   */
+  get<UK = K>(key: RelatedTo<K, UK>): V | undefined;
+  get<UK, O>(key: RelatedTo<K, UK>, otherwise: OptLazy<O>): V | O;
+  get<UK, O>(key: RelatedTo<K, UK>, otherwise?: OptLazy<O>): V | O | undefined {
+    const bestValue = getValueWithNearestKey(
+      this.context.distanceFunction,
+      key as K,
+      this.internalMap
+    );
+
+    return bestValue ?? OptLazy(otherwise);
```

Comment:
This works well in most cases, but the maps in rimbu are designed to deal with 'undefined' in the same way as any other value. So the following seems surprising but is logical:

```ts
const map = HashMap.of<string, number | undefined>(['a', undefined]);

const r1 = map.get('a');
// value of r1 will be undefined of course
const r2 = map.get('a', 5);
// value of r2 is still undefined, because the value was explicitly set
const r3 = map.get('b', 5);
// value of r3 is 5
```

This is why a lot of logic is using a local `token` symbol that is unique as fallback. You could the same technique in the `getValueWithNearestKey` function, so that it returns a symbol instead of undefined if there is no match.

This is a good reminder for me to add test cases for this kind of behavior in the test suites.

Reply:
🍻Thank you for the remark - I can clearly see your point, as from my perspective I was implicitly excluding the possibility of explicit `undefined` values. 🔎Revising the code - stay tuned! 🎉

Reply:
I've also added the test case about undefined values - right as you described above - to the generic map tests... and it now works for ProximityMap, too! 🥳

Reply:
Here we are! 🤠 In this batch of commits, I have:

* introduced the `get()` test case for generic maps which you described above
* revised the key matching algorithm - that now supports `undefined` values and also returns the closest key and the related distance in a structured way
* fixed `ProximityMap` to support the new code

Reply:
Additionally, in the latest commit I have registered the `proximity` package into `@rimbu/core`, just like the other collection types.

Reply:
Thanks for taking the effort! All looks good